src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -1412,13 +1418,17 @@ void handleAtomicBegin(CommandHandler * commandObj, const ConcreteCommandPath &
     // needs to keep track of a pending preset list now.
     delegate->InitializePendingPresets();
 
-    uint16_t maxTimeout = 5000;
-    timeout             = std::min(timeout, maxTimeout);
+    System::Clock::Milliseconds16 timeout =
+        delegate->GetAtomicWriteTimeout(commandData.attributeRequests, System::Clock::Milliseconds16(commandData.timeout.Value()));
 
-    ScheduleTimer(endpoint, System::Clock::Milliseconds16(timeout));
-    gThermostatAttrAccess.SetAtomicWrite(endpoint, true);
-    gThermostatAttrAccess.SetAtomicWriteScopedNodeId(endpoint, GetSourceScopedNodeId(commandObj));
-    sendAtomicResponse(commandObj, commandPath, imcode::Success, imcode::Success, imcode::Success, MakeOptional(timeout));
+    if (timeout == System::Clock::Milliseconds16(0))
```

Comment:
This error checking seems to lose the reason for the failure. Should we maybe make this CHIP_ERROR + out param or log inside GetAtomicWriteTimeout?

Reply:
I would also consider std::optional for the timeout, since "Timeout of 0 is invalid" feels a bit magic

Reply:
Switched to std::optional.